src/mmda/utils/stringify.py

```diff
@@ -0,0 +1,128 @@
+    document: Document,
+    use_word_text: Optional[bool] = True,
+    replace_newlines_with: Optional[str] = " ",
+    join_words_with: Optional[str] = " ",
+    normalize_whitespace: Optional[bool] = True,
+    allow_partial_word_match: Optional[bool] = True,
+    allow_disjoint_spans: Optional[bool] = True,
+    include_symbols_between_disjoint_spans: Optional[bool] = False,
+) -> str:
+    """Requires doc.words to exist. This technique applies 4 main steps:
+
+    1. Given a query span_group, find associated Words
+    2. Given associated Words, find associated Symbols (i.e. start-end indices)
+    3. Given associated Symbols, produce List[str] which includes whitespace chars
+    4. Given List[str], join into a single string
+    """
+    matched_words = document.find_overlapping(span_group, "words")
+
+    # warn if no words intersecting query span_group
+    if len(matched_words) == 0:
+        logger.warning(f"span_group {span_group} has no overlapping words in document")
+        return ""
+
+    # are words cut in half? suppose word is (0, 3), but the query asks for (2, 3). Should it include the whole word?
+    match_any_word_start = any(
+        [word.start == span_group.start for word in matched_words]
+    )
+    match_any_word_end = any([word.end == span_group.end for word in matched_words])
+    if allow_partial_word_match:
+        if not match_any_word_start:
+            logger.warning(
+                f"span_group {span_group}'s start index doesnt match w/ start of any word. output string may thus include more text than expected."
+            )
+        if not match_any_word_end:
+            logger.warning(
+                f"span_group {span_group}'s end index doesnt match w/ end of any word. output string may thus include more text than expected."
+            )
+    else:
+        if not match_any_word_start:
+            raise ValueError(
+                f"span_group {span_group}'s start index doesnt match w/ start of any word"
+            )
+        if not match_any_word_end:
+            raise ValueError(
+                f"span_group {span_group}'s end index doesnt match w/ end of any word"
+            )
+
+    # if query has disjoint spans, what should we do with the in-between symbols when stringifying?
+    if Span.are_disjoint(spans=span_group.spans):
+        if not allow_disjoint_spans:
+            raise ValueError(
+                f"span_group {span_group} has disjoint spans but allow_disjoint_spans is False"
+            )
+        else:
+            logger.warning(f"span_group {span_group} has disjoint spans")
+            if include_symbols_between_disjoint_spans:
+                # TODO: easiest is probably to convert disjoint spans into a single longer span
+                raise NotImplementedError
+
+    # if matched words are disjoint, what should we do with the in-between symbols when stringifying?
+    if Span.are_disjoint(spans=[span for word in matched_words for span in word.spans]):
```

Comment:
can you help me understand how this situation would arise? We were given SpanGroup spans that were overlapping, we fetched words associated with those spans, we're now checking the spans on the associated words, and they're disjoint? Which spans do we care about? Maybe that's what you mean with your comment on L97

Reply:
L84 is concerned w/ the case where, let's say an input query `SpanGroup` looks like:
```
[(0, 10), (15, 20)]
```
in this case, let's say it overlaps with the following words:
```
word1 -> (0, 10)
word2 -> (11, 14)
word3 -> (15, 20)
```
should the `stringify()` return the text from `word2` or just the text from `word1` and `word3`? If the latter, how should it stitch together?

I'm not sure when this would occur in SPP; it's sthn that @soldni  & I realized while implementing this that the intended behavior of how to handle `word2` isn't clear, so making it explicit here.